internal/provider/resource_gitlab_branch.go

```diff
@@ -0,0 +1,236 @@
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"committed_date": {
+			Description: "The date at which the commit was pushed.",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"committer_email": {
+			Description: "The email of the user that committed.",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"committer_name": {
+			Description: "The name of the user that committed.",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"short_id": {
+			Description: "The short id assigned to the commit by gitlab",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"title": {
+			Description: "The title of the commit",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"message": {
+			Description: "The commit message",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"parent_ids": {
+			Description: "The id of the parents of the commit",
+			Type:        schema.TypeSet,
+			Computed:    true,
+			Elem:        &schema.Schema{Type: schema.TypeString},
+			Set:         schema.HashString,
+		},
+	},
+}
+
+func resourceGitlabBranchCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	name := d.Get("name").(string)
+	project := d.Get("project").(string)
+	ref := d.Get("ref").(string)
+	branchOptions := &gitlab.CreateBranchOptions{
+		Branch: &name, Ref: &ref,
+	}
+
+	log.Printf("[DEBUG] create gitlab branch %s for project %s with ref %s", name, project, ref)
+	branch, resp, err := client.Branches.CreateBranch(project, branchOptions, gitlab.WithContext(ctx))
+	if err != nil {
+		log.Printf("[DEBUG] failed to create gitlab branch %v response %v", branch, resp)
+		return diag.FromErr(err)
+	}
+	d.SetId(buildTwoPartID(&project, &name))
+	return resourceGitlabBranchRead(ctx, d, meta)
+}
+
+func resourceGitlabBranchRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project, name, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[DEBUG] read gitlab branch %s", name)
+	branch, resp, err := client.Branches.GetBranch(project, name, gitlab.WithContext(ctx))
+	if err != nil {
+		if resp.StatusCode == 404 {
+			log.Printf("[DEBUG] recieved 404 for gitlab branch %s, removing from state", name)
+			d.SetId("")
+			return diag.FromErr(err)
+		}
+		log.Printf("[DEBUG] failed to read gitlab branch %s response %v", name, resp)
+		return diag.FromErr(err)
+	}
+	c := &gitlab.GetCommitRefsOptions{
+		Type: gitlab.String("branch"),
+	}
+	commitRefs, _, err := client.Commits.GetCommitRefs(project, branch.Commit.ID, c, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	ref := d.Get("ref").(string)
+	if ref == "" {
+		for _, br := range commitRefs {
+			if br.Name != branch.Name {
+				ref = br.Name
+				break
+			}
+		}
+	}
+	d.Set("ref", ref)
```

Comment:
What's the idea here? 

Can't we use the `commit.id` field as `ref` ?

Reply:
Nope then if the user defined a ref like "main" or "develop" it would be overwritten every time by the id from the commit and forcce replacement of the resources, one of the biggest issues with this feature is gitlab doesnt provide a consistent way to get the ref via the api but still requires this field for branch creation which I mentioned in some comments above -- also why originally my solution used ref with a default field rather than a required -- happy to revert back to this but this was also mentioned by another reviewer -- to me there is currently no perfect solution as the data source does not need the ref to find a branch otherwise I would suggest using something along the lines of a 3 part id and just using the defined ref from configuration rather than ever reading it from the api

Reply:
I see ... maybe we can also address this upstream.

In the meanwhile the currently implementation is okay for me. However, I think we can refactor it a little to only get `commitRefs` in case the `ref` is empty.

Reply:
Agreed, if any changes could be made on upstream that would be ideal but will implement this suggestion for now